### ISSUE TITLE
Facts for hosts in skipped plays are still gathered

### ISSUE THREAD

**Issue by hai-ld** [ansible/ansible#25412]
##### ISSUE TYPE

 - Bug Report

##### COMPONENT NAME

gather_facts

##### ANSIBLE VERSION

```
ansible 2.3.0.0
  config file = 
  configured module search path = Default w/o overrides
  python version = 2.7.12 (default, Jul  1 2016, 15:12:24) [GCC 5.4.0 20160609]
```

##### CONFIGURATION

Default configuration (no local ansible.cfg)

##### OS / ENVIRONMENT

OS = Ubuntu 16.04.2 LTS

##### SUMMARY

If a playbook consists of multiple play, each of which affects distinct set of hosts, when only a subset of play is selected for running, Ansible still tries to gather facts from unaffected hosts. This might be related to #21528

##### STEPS TO REPRODUCE

Run the following playbook `ansible-playbook -i dev sample.yml --tags=db`

```yaml
---

- hosts: mysql
  tags: db
  tasks:
    - debug: msg=hello

- hosts: app
  tags: app
  tasks:
    - debug: msg=world
```

##### EXPECTED RESULTS

Facts are not gathered for `app` hosts

##### ACTUAL RESULTS

Facts are gathered for all hosts

```
PLAY [mysql] ******************************************************************************************************************************************************************************************************

TASK [Gathering Facts] ********************************************************************************************************************************************************************************************
ok: [172.31.2.4]

TASK [debug] ******************************************************************************************************************************************************************************************************
ok: [172.31.2.4] => {
    "changed": false, 
    "msg": "hello"
}

PLAY [app] ********************************************************************************************************************************************************************************************************

TASK [Gathering Facts] ********************************************************************************************************************************************************************************************
ok: [172.31.1.4]

PLAY RECAP ********************************************************************************************************************************************************************************************************
172.31.1.4                : ok=1    changed=0    unreachable=0    failed=0   
172.31.2.4                 : ok=2    changed=0    unreachable=0    failed=0   

```

**Comment by bcoca** [comment]
fact gathering has an implicit 'always' tag, this was done to keep backwards compatibility with 1.x behaviour, which ignored tags.

**Comment by hai-ld** [comment]
@bcoca Is there a workaround, e.g. a flag/setting to disable the backward-compatible behaviour?

**Comment by nitzmahone** [comment]
Easiest is probably to explicitly set `gather_facts: no` and call the `setup` task yourself where you want it - then you have full control over when it does/doesn't run.

**Comment by hai-ld** [comment]
@nitzmahone Do Ansible team have any intention to deprecate this behaviour? I think it's quite unintuitive. My use case is having CI/CD server to automatically deploy new application version (on dev/QA env) but currently I can't put application deployment and environment setup (load balancers, database...) in the same playbook without also giving CI server access to those hosts, even though it has nothing to do with those.

**Comment by nitzmahone** [comment]
That'd be a question for @jimi-c, but I'd suspect the answer is no, especially since IIRC it's possible to use the result of fact gathering to determine whether a play is skipped, so that could be a breaking change.

**Comment by mkrizek** [comment]
Workaround mentioned in a duplicate issue https://github.com/ansible/ansible/issues/31569.

**Comment by bcoca** [comment]
fixed via #44717
